lib/full-site-editing/block-templates.php

```diff
@@ -0,0 +1,281 @@
+		$theme_template_files = _gutenberg_get_template_paths( $theme_dir . '/' . $template_base_paths[ $template_type ] );
+		foreach ( $theme_template_files as $template_file ) {
+			$template_base_path = $template_base_paths[ $template_type ];
+			$template_slug      = substr(
+				$template_file,
+				// Starting position of slug.
+				strpos( $template_file, $template_base_path . DIRECTORY_SEPARATOR ) + 1 + strlen( $template_base_path ),
+				// Subtract ending '.html'.
+				-5
+			);
+			$template_files[] = array(
+				'slug'  => $template_slug,
+				'path'  => $template_file,
+				'theme' => $theme_slug,
+				'type'  => $template_type,
+			);
+		}
+	}
+
+	return $template_files;
+}
+
+/**
+ * Build a unified template object based a theme file.
+ *
+ * @param array $template_file Theme file.
+ * @param array $template_type wp_template or wp_template_part.
+ *
+ * @return WP_Block_Template Template.
+ */
+function _gutenberg_build_template_result_from_file( $template_file, $template_type ) {
+	$default_template_types = gutenberg_get_default_template_types();
+
+	$theme               = wp_get_theme()->get_stylesheet();
+	$template            = new WP_Block_Template();
+	$template->id        = $theme . '|' . $template_file['slug'];
+	$template->theme     = $theme;
+	$template->content   = file_get_contents( $template_file['path'] );
+	$template->slug      = $template_file['slug'];
+	$template->is_custom = false;
+	$template->type      = $template_type;
+	$template->title     = $template_file['slug'];
+	$template->status    = 'publish';
+
+	if ( 'wp_template' === $template_type && isset( $default_template_types[ $template_file['slug'] ] ) ) {
+		$template->description = $default_template_types[ $template_file['slug'] ]['description'];
+		$template->title       = $default_template_types[ $template_file['slug'] ]['title'];
+	}
+
+	return $template;
+}
+
+/**
+ * Build a unified template object based a post Object.
+ *
+ * @param WP_Post $post Template post.
+ *
+ * @return WP_Block_Template Template.
+ */
+function _gutenberg_build_template_result_from_post( $post ) {
+	$theme = get_the_terms( $post, 'wp_theme' )[0]->slug;
+
+	$template              = new WP_Block_Template();
+	$template->wp_id       = $post->ID;
+	$template->id          = $theme . '|' . $post->post_name;
+	$template->theme       = $theme;
+	$template->content     = $post->post_content;
+	$template->slug        = $post->post_name;
+	$template->is_custom   = true;
+	$template->type        = $post->post_type;
+	$template->description = $post->post_excerpt;
+	$template->title       = $post->post_title;
+	$template->status      = $post->post_status;
+
+	return $template;
+}
+
+/**
+ * Retrieves a list of unified template objects based on a query.
+ *
+ * @param array $query Query.
+ * @param array $template_type wp_template or wp_template_part.
+ *
+ * @return array Templates.
+ */
+function gutenberg_get_block_templates( $query = array(), $template_type = 'wp_template' ) {
```

Comment:
This function has quite a bit of complex logic -- especially those `if` conditionals that check for the presence of different parts of `$query`. I think it'd be great if we could add unit tests for the specific scenarios that they're covering. Upon first reading, it's not totally easy to tell what those scenarios are, so unit tests would not only help to guard against regressions, but also to showcase/explain those scenarios.